compiler/src/iree/compiler/Dialect/Encoding/IR/test/roundtrip.mlir

```diff
@@ -37,40 +37,72 @@ func.func @set_encoding_with_batch_matmul_user(%arg0: tensor<?x?x?xf32>) {
+// CHECK-SAME: %[[D0:[a-zA-Z0-9]+]]
+// CHECK-SAME: %[[D1:[a-zA-Z0-9]+]]
+func.func @unset_encoding_fully_dynamic(%arg0: tensor<?x?xf32, #iree_encoding.encoding<operand_index = 1 : i64, op_type = matmul, element_types = [f32, f32, f32]>>,
+    %d0 : index, %d1 : index) -> tensor<?x?xf32> {
+  // CHECK: iree_encoding.unset_encoding %[[ARG0]] : tensor<?x?xf32, #iree_encoding.encoding<operand_index = 1 : i64, op_type = matmul, element_types = [f32, f32, f32]>> -> tensor<?x?xf32>{%[[D0]], %[[D1]]}
+  %0 = iree_encoding.unset_encoding %arg0 : tensor<?x?xf32, #iree_encoding.encoding<operand_index = 1 : i64, op_type = matmul, element_types = [f32, f32, f32]>> -> tensor<?x?xf32>{%d0, %d1}
   return %0 : tensor<?x?xf32>
 }
 
 // -----
 
-// CHECK: @unset_encoding_ops_mixed_dynamic_static(%[[ARG0:.+]]: tensor<10x?xf32, #iree_encoding.encoding<operand_index = 1 : i64, op_type = matmul, element_types = [f32, f32, f32]>>)
-func.func @unset_encoding_ops_mixed_dynamic_static(%arg0: tensor<10x?xf32, #iree_encoding.encoding<operand_index = 1, op_type = matmul, element_types = [f32, f32, f32]>>) -> tensor<?x20xf32> {
-  // CHECK: iree_encoding.unset_encoding %[[ARG0]] : tensor<10x?xf32, #iree_encoding.encoding<operand_index = 1 : i64, op_type = matmul, element_types = [f32, f32, f32]>>
-  %0 = iree_encoding.unset_encoding %arg0 : tensor<10x?xf32, #iree_encoding.encoding<operand_index = 1, op_type = matmul, element_types = [f32, f32, f32]>> -> tensor<?x20xf32>
+// CHECK: @unset_encoding_ops_mixed_dynamic_static(%[[ARG0:.+]]: tensor<10x?xf32, #iree_encoding.encoding<operand_index = 1 : i64, op_type = matmul, element_types = [f32, f32, f32]>>
+// CHECK-SAME: %[[D0:[a-zA-Z0-9]+]]
+func.func @unset_encoding_ops_mixed_dynamic_static(%arg0: tensor<10x?xf32, #iree_encoding.encoding<operand_index = 1, op_type = matmul, element_types = [f32, f32, f32]>>,
+    %d0 : index) -> tensor<?x20xf32> {
+  // CHECK: iree_encoding.unset_encoding %[[ARG0]] : tensor<10x?xf32, #iree_encoding.encoding<operand_index = 1 : i64, op_type = matmul, element_types = [f32, f32, f32]>> -> tensor<?x20xf32>{%[[D0]]}
+  %0 = iree_encoding.unset_encoding %arg0 : tensor<10x?xf32, #iree_encoding.encoding<operand_index = 1, op_type = matmul, element_types = [f32, f32, f32]>> -> tensor<?x20xf32>{%d0}
   return %0 : tensor<?x20xf32>
 }
 
 // -----
 
+// CHECK: @unset_encoding_op_with_extract_slice_change(
+// CHECK-SAME: %[[ARG0:[a-zA-Z0-9]+]]
+#encoding = #iree_encoding.encoding<operand_index = 0, op_type = matmul, element_types = [f32, f32, f32]>
+func.func @unset_encoding_op_with_extract_slice_change(%arg0 : tensor<20x30xf32, #encoding>) -> tensor<10x20xf32> {
+  // CHECK: iree_encoding.unset_encoding %[[ARG0]] : tensor<20x30xf32, #iree_encoding.encoding<operand_index = 0 : i64, op_type = matmul, element_types = [f32, f32, f32]>> -> tensor<10x20xf32>
+  %0 = iree_encoding.unset_encoding %arg0: tensor<20x30xf32, #encoding> -> tensor<10x20xf32>
+  return %0 : tensor<10x20xf32>
```

Comment:
Is this actually valid? I thought the result shape of the unset_encoding op should match the source shape.

Reply:
THat is what is being changed here? This adds slicing semantics to `unset_encoding`.

Reply:
What Mahesh said. This is the idea of the PR.

Reply:
Maybe this depends on how we define the semantics of set_encoding. My understanding has always been that the set_encoding op should not change the shape of the source tensor, but only add an encoding attribute to it. This would mean that the unset_encoding op also should not change the shape of the source tensor, which would mean this unset_encoding op is not valid.

My understanding was that this PR will only affect dynamic sizes, since we cannot recover the original sizes from just the tensor type when it is time to materialize encodings without carrying the dynamic dimensions. When the dimensions are static, we have the sizes in the type. This does not change the fact that when it is time to set encodings, we do not change the shape of the encoded tensors.

In fact, I think there is good reason not to change the type of the encoded tensors. Changing the shape of an encoded tensor already adds some explicit padding to the tensor before materialization. This means that any transformation on the encoded tensors that happens before materialization will be using information that may change after materializing the encoding into nop. This could potentially introduce similar bugs to what this PR is trying to fix, where an encoding has an affect on the program even when the encoding materializes into a nop.

Reply:
We had an offline discussion this morning, and I'm convinced. Note that (1) this does not impact any data-tiling project and it is more like discussing the op semantics, and (2) it is adding more restriction to the ops which should not hurt anything and we can revisit it when there is a need.

The main point is that we want to make set_encoding and unset_encoding consistent. We have padding semantics in set_encoding, but we're not able to fold tensor.pad ops into the encoding ops. Because all the information can't get resolved until we have target arch information. With this change, it's possible to fold extract_slice op into unset_encoding ops, which looks more powerful than set_encoding ops. It looks inconsistent to me between set_encoding ops and unset_encoding ops. Thus, I'm convinced that we can have the check for this case.

Repeating the points again: this test case is not used at all in IREE, and it is more like exploring the actual op semantics.